File: frontend/src/types/workspace.ts
/**
 * Workspace wire contracts — mirror backend/app/models/workspace.py exactly
 * (separate module: never extend types/models.ts). All records are persistent
 * declarations written only on explicit CRUD; PlacementInspection is a runtime
 * observation that lives in RAM on both sides and is never persisted.
 */

export type ArtifactKind = "code" | "dataset" | "model";

export type InspectionState = "declared" | "verified" | "missing" | "unavailable";

export interface ProjectRecord {
  project_id: string;
  name: string;
  description: string;
  artifact_ids: string[];
  launch_config_ids: string[];
  tags: string[];
  /** Default transfer exclusion patterns (rsync/fnmatch) for this project. */
  transfer_excludes: string[];
  created_at: string;
  updated_at: string;
}

export interface ProjectCreate {
  name: string;
  description?: string;
  artifact_ids?: string[];
  tags?: string[];
}

export interface ProjectPatch {
  name?: string;
  description?: string;
  artifact_ids?: string[];
  tags?: string[];
  transfer_excludes?: string[];
}

export interface ArtifactRecord {
  artifact_id: string;
  kind: ArtifactKind;
  name: string;
  /** null = unversioned; content changes create a new version instead. */
  version: string | null;
  description: string;
  immutable: boolean;
  created_at: string;
  updated_at: string;
}

export interface ArtifactCreate {
  kind: ArtifactKind;
  name: string;
  version?: string | null;
  description?: string;
  /** Omitted → backend default: immutable unless kind is code. */
  immutable?: boolean | null;
}

export interface ArtifactPatch {
  name?: string;
  version?: string | null;
  description?: string;
  immutable?: boolean | null;
}

export interface PlacementRecord {
  placement_id: string;
  artifact_id: string;
  server_id: string;
  remote_path: string;
  created_at: string;
  updated_at: string;
}

export interface PlacementCreate {
  artifact_id: string;
  server_id: string;
  remote_path: string;
}

export interface PlacementPatch {
  remote_path?: string;
}

/** RAM-only scan result; `checked_at` marks when the explicit inspect ran. */
export interface PlacementInspection {
  placement_id: string;
  state: InspectionState;
  file_type: string | null;
  size_b: number | null;
  file_count: number | null;
  checked_at: string;
  detail: string;
}

export interface LaunchConfigRecord {
  launch_config_id: string;
  project_id: string;
  name: string;
  working_dir: string | null;
  /** Structured launch: program + args[], never a shell string. */
  program: string;
  args: string[];
  environment: string | null;
  env_vars: Record<string, string>;
  required_artifact_ids: string[];
  gpu_count: number | null;
  min_vram_b: number | null;
  created_at: string;
  updated_at: string;
}

export interface LaunchConfigCreate {
  project_id: string;
  name: string;
  working_dir?: string | null;
  program: string;
  args?: string[];
  environment?: string | null;
  env_vars?: Record<string, string>;
  required_artifact_ids?: string[];
  gpu_count?: number | null;
  min_vram_b?: number | null;
}

export interface LaunchConfigPatch {
  name?: string;
  working_dir?: string | null;
  program?: string;
  args?: string[];
  environment?: string | null;
  env_vars?: Record<string, string>;
  required_artifact_ids?: string[];
  gpu_count?: number | null;
  min_vram_b?: number | null;
}

export interface ServerRoots {
  project_root: string | null;
  dataset_root: string | null;
  model_root: string | null;
  output_root: string | null;
}

/** PUT replaces the whole record: omit → null (cleared). Fields are 1–512
 *  chars when present, so an empty input is sent as null, never "". */
export interface ServerRootsUpdate {
  project_root?: string | null;
  dataset_root?: string | null;
  model_root?: string | null;
  output_root?: string | null;
}

File: backend/app/models/workspace.py
"""Workspace wire models (Sol-owned contract).

Persistent declarations only; runtime observations (scan results, transfer
progress) live in RAM and are modelled separately. `extra="forbid"` everywhere.
"""

from __future__ import annotations

from enum import StrEnum

from pydantic import BaseModel, ConfigDict, Field, field_validator

_SHELL_META = set(";&|<>`$\n\r\x00(){}[]*?~\"'\\\\")


class ArtifactKind(StrEnum):
    CODE = "code"
    DATASET = "dataset"
    MODEL = "model"


_ID = Field(min_length=1, max_length=64)
_NAME = Field(min_length=1, max_length=120)
_DESC = Field(default="", max_length=600)


class ProjectRecord(BaseModel):
    model_config = ConfigDict(extra="forbid")

    project_id: str = _ID
    name: str = _NAME
    description: str = _DESC
    artifact_ids: list[str] = Field(default_factory=list, max_length=64)
    launch_config_ids: list[str] = Field(default_factory=list, max_length=32)
    tags: list[str] = Field(default_factory=list, max_length=16)
    created_at: str = ""
    updated_at: str = ""


class ProjectCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str = _NAME
    description: str = _DESC
    artifact_ids: list[str] = Field(default_factory=list, max_length=64)
    tags: list[str] = Field(default_factory=list, max_length=16)


class ProjectPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=600)
    artifact_ids: list[str] | None = Field(default=None, max_length=64)
    tags: list[str] | None = Field(default=None, max_length=16)


class ArtifactRecord(BaseModel):
    model_config = ConfigDict(extra="forbid")

    artifact_id: str = _ID
    kind: ArtifactKind
    name: str = _NAME
    version: str | None = Field(default=None, max_length=64)
    description: str = _DESC
    immutable: bool = True
    created_at: str = ""
    updated_at: str = ""


class ArtifactCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    kind: ArtifactKind
    name: str = _NAME
    version: str | None = Field(default=None, max_length=64)
    description: str = _DESC
    immutable: bool | None = None  # default: immutable unless kind == code


class ArtifactPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=120)
    version: str | None = Field(default=None, max_length=64)
    description: str | None = Field(default=None, max_length=600)
    immutable: bool | None = None


_PATH = Field(min_length=1, max_length=512)


class PlacementRecord(BaseModel):
    model_config = ConfigDict(extra="forbid")

    placement_id: str = _ID
    artifact_id: str = _ID
    server_id: str = _ID
    remote_path: str = _PATH
    created_at: str = ""
    updated_at: str = ""


class PlacementCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    artifact_id: str = _ID
    server_id: str = _ID
    remote_path: str = _PATH


class PlacementPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    remote_path: str | None = _PATH


class InspectionState(StrEnum):
    DECLARED = "declared"
    VERIFIED = "verified"
    MISSING = "missing"
    UNAVAILABLE = "unavailable"


class PlacementInspection(BaseModel):
    """Runtime observation — RAM only, never persisted."""

    model_config = ConfigDict(extra="forbid")

    placement_id: str = _ID
    state: InspectionState
    file_type: str | None = None  # directory | regular file | ...
    size_b: int | None = None
    file_count: int | None = None
    checked_at: str = ""
    detail: str = ""


class LaunchConfigRecord(BaseModel):
    model_config = ConfigDict(extra="forbid")

    @field_validator("program", "working_dir")
    @classmethod
    def _no_shell_metacharacters(cls, value: str | None) -> str | None:
        if value is not None:
            for char in _SHELL_META:
                if char in value:
                    raise ValueError(
                        "shell metacharacters are not allowed; use structured program + args"
                    )
        return value

    launch_config_id: str = _ID
    project_id: str = _ID
    name: str = _NAME
    working_dir: str | None = Field(default=None, min_length=1, max_length=512)
    program: str = Field(min_length=1, max_length=256)
    args: list[str] = Field(default_factory=list, max_length=64)
    environment: str | None = Field(default=None, max_length=120)
    env_vars: dict[str, str] = Field(default_factory=dict, max_length=32)
    required_artifact_ids: list[str] = Field(default_factory=list, max_length=64)
    gpu_count: int | None = Field(default=None, ge=1, le=64)
    min_vram_b: int | None = Field(default=None, ge=0)
    created_at: str = ""
    updated_at: str = ""


class LaunchConfigCreate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    @field_validator("program", "working_dir")
    @classmethod
    def _no_shell_metacharacters(cls, value: str | None) -> str | None:
        if value is not None:
            for char in _SHELL_META:
                if char in value:
                    raise ValueError("shell metacharacters are not allowed; use structured args")
        return value

    project_id: str = _ID
    name: str = _NAME
    working_dir: str | None = Field(default=None, min_length=1, max_length=512)
    program: str = Field(min_length=1, max_length=256)
    args: list[str] = Field(default_factory=list, max_length=64)
    environment: str | None = Field(default=None, max_length=120)
    env_vars: dict[str, str] = Field(default_factory=dict, max_length=32)
    required_artifact_ids: list[str] = Field(default_factory=list, max_length=64)
    gpu_count: int | None = Field(default=None, ge=1, le=64)
    min_vram_b: int | None = Field(default=None, ge=0)


class LaunchConfigPatch(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str | None = Field(default=None, min_length=1, max_length=120)
    working_dir: str | None = _PATH
    program: str | None = Field(default=None, min_length=1, max_length=256)
    args: list[str] | None = Field(default=None, max_length=64)
    environment: str | None = Field(default=None, max_length=120)
    env_vars: dict[str, str] | None = Field(default=None, max_length=32)
    required_artifact_ids: list[str] | None = Field(default=None, max_length=64)
    gpu_count: int | None = Field(default=None, ge=1, le=64)
    min_vram_b: int | None = Field(default=None, ge=0)


class ServerRoots(BaseModel):
    model_config = ConfigDict(extra="forbid")

    project_root: str | None = Field(default=None, min_length=1, max_length=512)
    dataset_root: str | None = Field(default=None, min_length=1, max_length=512)
    model_root: str | None = Field(default=None, min_length=1, max_length=512)
    output_root: str | None = Field(default=None, min_length=1, max_length=512)


class ServerRootsUpdate(BaseModel):
    model_config = ConfigDict(extra="forbid")

    project_root: str | None = Field(default=None, min_length=1, max_length=512)
    dataset_root: str | None = Field(default=None, min_length=1, max_length=512)
    model_root: str | None = Field(default=None, min_length=1, max_length=512)
    output_root: str | None = Field(default=None, min_length=1, max_length=512)
